nightingale_manipulation/src/nightingale_manipulation/manipulation_action_client.py

```diff
@@ -132,55 +299,215 @@ def __init__(self):
+        time.sleep(1)
+        if self.update_ee_pose():
+            rospy.loginfo(f"{prefix} arm cartesian control initialized")
+        else:
+            rospy.logerr(f"{prefix} arm failed to initialize")
 
-        self.left_arm = actionlib.simple_action_client.SimpleActionClient(
-            "/moveit_action_handlers/left_arm/cartesian_ctrl", MoveToPoseMoveItAction
-        )
-        self.left_arm.wait_for_server()
+    def update_ee_pose(self):
+        """
+        Uses the robot transform to update the end effector Pose and Orientation
+
+        @return: True if successful
+        """
+        try:
+            link_transform = self.tf_buffer.lookup_transform(
+                self.ref_link, self.ee_link, rospy.Time(0)
+            )
+        except Exception as e:
+            rospy.logerr(str(e))
+            rospy.logerr(f"{self.prefix} arm transformation lookup failed")
+            return False
+
+        q = link_transform.transform.rotation
+
+        quaternion_list = [q.x, q.y, q.z, q.w]
+        (roll, pitch, yaw) = euler_from_quaternion(quaternion_list)
+        self.orientation.roll = roll
+        self.orientation.pitch = pitch
+        self.orientation.yaw = yaw
+
+        p = link_transform.transform.translation
+        self.pose.position = p
+        self.pose.orientation = q
+        return True
+
+    def cmd_arm(self, pose: GeometryPose, blocking=True, ee_relative=False):
+        """
+        Commands the arm to move to a target Pose in the cartesian space
+
+        @param pose: type geometry_msgs.msg Pose, also accepts manipulation_action_client.Pose type
+        @param blocking: if true, wait for action server result
+        @param ee_relative: if true, use ee_link instead of base_link for reference link
+        @return: Result of action
+        """
+        if not self.update_ee_pose():
+            return False
+        if self.pose == pose:
+            return True
+
+        if ee_relative:
+            ref_link = self.ee_link
+        else:
+            ref_link = self.ref_link
+
+        # locked end effector control mode
+        if self.ee_ctrl_mode:
+            goal = cartesian_goal(
+                pose.position, ref_link, self.orientation, mode=self.ee_ctrl_mode
+            )
+
+        # free end effector control mode
+        else:
+            quaternion_list = [
+                pose.orientation.x,
+                pose.orientation.y,
+                pose.orientation.z,
+                pose.orientation.w,
+            ]
+            (roll, pitch, yaw) = euler_from_quaternion(quaternion_list)
+            euler_orientation = Orientation(roll=roll, pitch=pitch, yaw=yaw)
+            goal = cartesian_goal(pose.position, ref_link, euler_orientation)
+
+        print("goal: ")
+        print(goal)
+
+        self.arm.send_goal(goal)
+        if blocking:
+            self.arm.wait_for_result()
+            return self.arm.get_result()
```

Comment:
can this return a bool? so check if get_result() is good before u return? the caller wont know what the meaning of the get_result() is (unless get_result() already gives u a bool)